pkg/shares/share_splitting.go

```diff
@@ -1,412 +1,120 @@
+	// that do not follow the non-interactive defaults
+	msgIndexes := ExtractShareIndexes(data.Txs)
 
-	for i := 0; i < consts.ShareReservedBytes; i++ {
-		// here we force the last share reserved byte to be zero to avoid any
-		// confusion for light clients parsing these shares, as the rest of the
-		// data after transaction is padding. See
-		// https://github.com/celestiaorg/celestia-specs/blob/master/src/specs/data_structures.md#share
-		rawLastShare[consts.NamespaceSize+i] = byte(0)
+	var msgShares [][]byte
+	if msgIndexes != nil && int(msgIndexes[0]) != currentShareCount {
+		return nil, ErrUnexpectedFirstMessageShareIndex
 	}
 
-	newLastShare := NamespacedShare{
-		Share: rawLastShare,
-		ID:    lastShare.NamespaceID(),
+	msgShares, err = SplitMessages(msgIndexes, data.Messages.MessagesList)
+	if err != nil {
+		return nil, err
 	}
-	csw.shares[len(csw.shares)-1] = newLastShare
-	return csw.shares
-}
+	currentShareCount += len(msgShares)
 
-// Count returns the current number of shares that will be made if exporting.
-func (csw *ContiguousShareWriter) Count() (count, availableBytes int) {
-	availableBytes = consts.TxShareSize - (len(csw.pendingShare.Share) - consts.NamespaceSize)
-	return len(csw.shares), availableBytes
-}
+	tailShares := TailPaddingShares(wantShareCount - currentShareCount).RawShares()
 
-// tail is filler for all tail padded shares
-// it is allocated once and used everywhere
-var tailPaddingShare = append(
-	append(make([]byte, 0, consts.ShareSize), consts.TailPaddingNamespaceID...),
-	bytes.Repeat([]byte{0}, consts.ShareSize-consts.NamespaceSize)...,
-)
+	// todo: optimize using a predefined slice
+	shares := append(append(append(
+		txShares,
+		evdShares...),
+		msgShares...),
+		tailShares...)
 
-func TailPaddingShares(n int) NamespacedShares {
-	shares := make([]NamespacedShare, n)
-	for i := 0; i < n; i++ {
-		shares[i] = NamespacedShare{
-			Share: tailPaddingShare,
-			ID:    consts.TailPaddingNamespaceID,
+	return shares, nil
+}
+
+// ExtractShareIndexes iterates over the transactions and extracts the share
+// indexes from wrapped transactions. It returns nil if the transactions are
+// from an old block that did not have share indexes in the wrapped txs.
+func ExtractShareIndexes(txs coretypes.Txs) []uint32 {
+	var msgIndexes []uint32
+	for _, rawTx := range txs {
+		if malleatedTx, isMalleated := coretypes.UnwrapMalleatedTx(rawTx); isMalleated {
+			// Since share index == 0 is invalid, it indicates that we are
+			// attempting to extract share indexes from txs that do not have any
+			// due to them being old. here we return nil to indicate that we are
+			// attempting to extract indexes from a block that doesn't support
+			// it. It's check for 0 because if there is a message in the block,
+			// then there must also be a tx, which will take up at least one
+			// share.
+			if malleatedTx.ShareIndex == 0 {
+				return nil
+			}
+			msgIndexes = append(msgIndexes, malleatedTx.ShareIndex)
```

Comment:
here we make this code compatible with blocks of transactions that don't use wrapped txs or specify the share index. This will be useful if we want to preserve the ability of syncing from scratch on mamaki.